### PR TITLE
sync : ggml (new ops, tests, backend, etc.)

### PR DESCRIPTION
**Pull request by ggerganov**
Careful look and testing required - very likely to have broken something during this merge

### CODE REVIEW

**Comment by ggerganov** [comment]
Need help with enabling CUDA separable compilation in the Makefile

**Comment by slaren** [comment]
I think we can ignore that for now, that was used for the device-side mul mat id implementation, but that path has been disabled for now. The kernel that requires dynamic parallelism should be in an `#if 0` block.

**Comment by ggerganov** [comment]
I see - it seems still required for the dequantization kernels:

```
ggml-cuda.cu(5229): error: kernel launch from __device__ or __global__ functions requires separate compilation mode
      dequantize_block<qk, qr, dequantize_kernel><<<num_blocks, 256, 0, stream>>>(vx, y, k);
      ^

ggml-cuda.cu(5236): error: kernel launch from __device__ or __global__ functions requires separate compilation mode
      dequantize_block_q2_K<<<nb, 64, 0, stream>>>(vx, y);
      ^

ggml-cuda.cu(5246): error: kernel launch from __device__ or __global__ functions requires separate compilation mode
      dequantize_block_q3_K<<<nb, 64, 0, stream>>>(vx, y);
      ^

ggml-cuda.cu(5255): error: kernel launch from __device__ or __global__ functions requires separate compilation mode
      dequantize_block_q4_K<<<nb, 32, 0, stream>>>(vx, y);
      ^

ggml-cuda.cu(5262): error: kernel launch from __device__ or __global__ functions requires separate compilation mode
      dequantize_block_q5_K<<<nb, 64, 0, stream>>>(vx, y);
      ^

ggml-cuda.cu(5272): error: kernel launch from __device__ or __global__ functions requires separate compilation mode
      dequantize_block_q6_K<<<nb, 64, 0, stream>>>(vx, y);
      ^

6 errors detected in the compilation of "ggml-cuda.cu".
make: *** [Makefile:451: ggml-cuda.o] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by slaren** [comment]
That should fix it.

File: ggml-cuda.cu [hunk 1]
```diff
@@ -6903,6 +7044,7 @@ inline void ggml_cuda_op_rope(
         GGML_ASSERT(false);
         rope_glm_f32_cuda(src0_dd, dst_dd, ne00, nrows, pos, freq_scale, ne01, freq_base, n_ctx, main_stream);
     } else if (is_neox) {
+        GGML_ASSERT(ne00 == n_dims && "ne00 != n_dims is not implemented for CUDA yet");
```
Comment (slaren) [inline ggml-cuda.cu:7047]: This assert also needs to be removed.

**Comment by slaren** [comment]
Are you going to sync the tests too? It's in the PR title, but not in the files.

**Comment by ggerganov** [comment]
Yes, I forgot - will sync them in a few

**Comment by ggerganov** [comment]
Seems like when linking to `libllama` the backend registers do not work. Happens with the `cmake` build. I think it needs some "whole-archive" flags, but I've forgotten how this works

**Comment by slaren** [comment]
Ok, I was afraid that this way of self-registering backends was going to be more trouble than it is worth. It may be better to scratch it and just have a list of backends in the registry depending on the compilation flags.

**Comment by ggerganov** [comment]
Yeah, my experience with such kind of logic is also not very positive - various platform compatibility problems arise. List of backends based on flags should work, despite the extra manual work to maintain the list

**Pull request merged**